Tweak color picker, angularize legend

rashidkpc opened the pull request:
This adds a few features and cleanups:

- Moved legend out to an angular directive
- Added more colors
- Added the ability to filter values in or out (as opposed to old "just in" behavior)
- Persisted the state of the legend (open/closed)
- Tweaked the legend collapse look

Tried to upload a screenshot, but Github is having a bad night

rashidkpc commented:
I also didn't rebase this on your branch, and I did merge master...so there's that :-D

lukasolson commented:
This is looking amazing. I noticed one bug, when you have multiple visualizations on a dashboard, then click on one of the legend values, it toggles the color selection/filtering, but then toggles it again really quick so you can't do anything.

lukasolson commented:
Also, if this could be hidden until the content actually shows up, that would be good. It's kind of startling.

![image](https://cloud.githubusercontent.com/assets/1178348/11576285/ccda1dce-99d2-11e5-8cf8-e90c3c178ee2.png)

rashidkpc commented:
There's a few issues with this, not the least of which is that it leaks DOM nodes. The scopes are being cleaned up correctly, but something is still holding on. Investigating.

The disappearing thing is due to an issue in which the chart is short enough that opening the color picker causes a scrollbar to be drawn, which pushes the legend over, which makes the chart smaller, which causes a resize event, which causes the legend to be redrawn. IMO the correct way to fix that is to keep the legend from being redrawn. Going to have a go at it.

rashidkpc commented:
Currently tests are broken, will fix. This now angularizes the entire legend, completely removing responsibility for render it from the vislib

rashidkpc commented on src/ui/public/visualize/visualize_legend.js:
```diff
@@ -0,0 +1,102 @@
+          //return filters == null ? false : true;
+        };
+
+        $scope.colors = [
```
This should move out to another file

rashidkpc commented on src/ui/public/visualize/visualize.js:
```diff
@@ -47,19 +50,16 @@ define(function (require) {
           var fullSpy = ($scope.spy.mode && ($scope.spy.mode.fill || $scope.fullScreenSpy));
 
-          // external
-          $el.toggleClass('only-visualization', !$scope.spy.mode);
```
This was only used for making the spy take up the full container, and was sort of confusing in how it worked. The new method just looks to see if the resulting chart container is < 100px tall.

rashidkpc replied:
Note to self: Make sure full screen button works

rashidkpc replied:
Check, works

rashidkpc commented on src/ui/public/vislib/lib/dispatch.js:
```diff
@@ -226,16 +226,8 @@ define(function (require) {
-
       if (!label) return;
-
-      d3.select(element)
```
Since the legend isn't part of the Vis object now, rather the visualize directive, we need to get at the parent to make this work. Css selectors and Jquery make this a bit more concise

rashidkpc commented on src/ui/public/utils/find_by_param.js:
```diff
@@ -0,0 +1,13 @@
+define(function (require) {
```
Pulled this out of `src/ui/public/filter_bar/filter_bar_click_handler.js` into a `utils/` function so it could be reused.

rashidkpc commented on src/ui/public/filter_bar/filter_bar_click_handler.js:
```diff
@@ -2,21 +2,11 @@ define(function (require) {
   return function (Notifier) {
     return function ($state) {
-      return function (event) {
+      return function (event, simulate) {
```
Added `simulate` so you can check whether or not a filter would be created by this function. Allows us to check if we should display the filter buttons without reimplementing all of the logic in this file somewhere else.

rashidkpc commented:
Memory timeline, no growth in memory/node/listeners.

![screen shot 2015-12-07 at 1 52 44 pm](https://cloud.githubusercontent.com/assets/1250387/11639412/e7f1eb18-9ce9-11e5-9707-5f7886f6cb5f.png)

rashidkpc commented:
This is broken in the metric vis, adds a legend to it and shouldn't. Should probably check if its broken elsewhere

lukasolson commented:
![image](https://cloud.githubusercontent.com/assets/1178348/11666533/40d35ba2-9daa-11e5-87e5-036552e4fe54.png)
Looks like it's collapsed by default on the discover view, which actually, now that I think about it, might have been how it was before. The problem, however is, that you have to click it twice to get it to expand.

lukasolson commented:
I don't know if it was intended to have similar colors grouped in the same column in the color picker, but when you have a scrollbar (or on the dashboard) the colors aren't laid out the same:

![image](https://cloud.githubusercontent.com/assets/1178348/11667306/2379f77e-9dae-11e5-9af8-a570d689d079.png)

Wouldn't consider this a bug though.

lukasolson commented:
I think I might prefer having the filter buttons closer to the top rather than at the bottom, just because if you're in a small area (thinking dashboard here) you sometimes have to scroll down to get to the filter buttons. What do you think?

lukasolson commented:
When you add multiple metrics, then switch to percentage mode, the colors all change (because the labels have changed). I don't know if this is scoped to this initial issue or not, but it's a funky behavior.

![image](https://cloud.githubusercontent.com/assets/1178348/11669886/37cb94a4-9dbc-11e5-8cd6-c0977b13fdbf.png)

![image](https://cloud.githubusercontent.com/assets/1178348/11669895/42c84f8c-9dbc-11e5-85df-c5a831212c77.png)

lukasolson commented:
Other than that, LGTM. I don't see the issue you're talking about with the metric vis...

rashidkpc commented:
Moved the labels up, made the width of the color picker static, fixed the default legend state. 

As far as colors changing, its because the color is directly tied to the displayed string. It will be a major undertaking to divorce that model/view conflict. I'd say outside of the scope of this.

lukasolson commented:
Nitpick, but it looks kinda strange on discover having this button just kinda float out there, unlike on the visualize and dashboard screens where the toggle button sits flush with something on its right.

![image](https://cloud.githubusercontent.com/assets/1178348/11746534/c6007e4e-9fdb-11e5-9627-e995238d3ee3.png)

![image](https://cloud.githubusercontent.com/assets/1178348/11746580/ee96b94a-9fdb-11e5-8deb-4dbbc7dbec61.png)

![image](https://cloud.githubusercontent.com/assets/1178348/11746588/f4e74f94-9fdb-11e5-839e-c3c30180101b.png)

lukasolson commented:
LGTM!
